src/main/scala/com/fasterxml/jackson/module/scala/deser/OptionDeserializerModule.scala

```diff
@@ -68,7 +68,8 @@ private class OptionDeserializer(fullType: JavaType,
     if (t == JsonToken.VALUE_NULL) {
       getNullValue(ctxt)
     } else {
-      typeDeserializer.deserializeTypedFromAny(jp, ctxt).asInstanceOf[Option[AnyRef]]
+      val value = valueTypeDeserializer.get.deserializeTypedFromAny(jp, ctxt)
```

Comment:
This doesn't quite seem right. I believe that this `deserializeWithType` method is supposed to use the passed in `typeDeserializer` as an override. I will have to look-through/debug your example/test-case more carefully to see if I can figure out how/why it's passing a deserializer that does not work.